udn: set persistentIPs for UDN ifaces

maiqueb opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

All changes must adhere to this template to make it easy for reviewers
and preserve rationale/history behind every change
-->

#### What this PR does and why is it needed
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->
This PR looks for a special annotation on pods, where KubeVirt (or an agent on behalf of KubeVirt) will indicate what is the name of the `IPAMClaim` used for the primary UDN interface.

An alternative would be to use an hard-coded value, but then we wouldn't be able to distinguish between a pod and a VM, and we do not want to abuse reading the KubeVirt label.

#### Which issue(s) this PR fixes
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->
Pending adding the CI requirements, ensuring kubevirt e2e tests are run.

#### How to verify it
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->

#### Details to documentation updates
<!--
Did you include good docs that explain to our end users/developers/contributors
how your code works?
-->


#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Configure persistent IPs on UDN networks when requested

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: TBD
-->
```release-note
Configure persistent IPs on UDN networks when requested
```

maiqueb commented on go-controller/pkg/util/pod_annotation.go:
```diff
@@ -51,6 +51,9 @@ const (
 	DefNetworkAnnotation = "v1.multus-cni.io/default-network"
+	// OvnUDNIPAMClaimName is used for workload owners to instruct OVN-K which
+	// IPAMClaim will hold the allocation for the workload
+	OvnUDNIPAMClaimName = "k8s.ovn.org/ovn-udn-ipamclaim-reference"
```
@trozet / @tssurya 

let's discuss the annotation value.

jcaamano replied:
This could just be `k8s.ovn.org/ipamclaim-reference`

maiqueb replied:
really ? I think it should somehow indicate this is meant to tell ovn-k of what is the claim to be used for the primary UDN. 

I agree dropping the `ovn` should happen ... awfully redundant.

maiqueb replied:
Done.

maiqueb replied:
@qinqon we'll need to update the client for the e2e tests to pass ...

jcaamano commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -939,6 +939,13 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *kapi.Pod, nInfo NetInfo, ac
 	}
 
+	if nInfo.IsPrimaryNetwork() && nInfo.TopologyType() == types.Layer2Topology && nInfo.AllowsPersistentIPs() {
+		ipamClaimName, wasPersistentIPRequested := pod.Annotations[OvnUDNIPAMClaimName]
+		if wasPersistentIPRequested {
+			networkSelections[activeNetworkNADs[0]].IPAMClaimReference = ipamClaimName
+		}
```
nit: wouldn't just this be equivalent:
```
networkSelections[activeNetworkNADs[0]].IPAMClaimReference = pod.Annotations[OvnUDNIPAMClaimName]
```

maiqueb replied:
maybe. I try to repeat what I think are best practices without thinking too much. 
Always checking if something was found seems good. 

In this particular case I think you're right.

jcaamano commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -939,6 +939,13 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *kapi.Pod, nInfo NetInfo, ac
 		Name:      activeNetworkNADKey[1],
 	}
 
+	if nInfo.IsPrimaryNetwork() && nInfo.TopologyType() == types.Layer2Topology && nInfo.AllowsPersistentIPs() {
```
it would probably be useful to have a free `AllowsPersistentIPs(netInfo)` method here that summarizes which network allow persistent IPs

maiqueb replied:
OK, I'll add it.

maiqueb replied:
Done.

jcaamano commented on test/e2e/kubevirt.go:
```diff
@@ -313,16 +315,31 @@ var _ = Describe("Kubevirt Virtual Machines", func() {
-		httpServerTestPodsMultusNetworkIPs = func(netName string) map[string][]string {
+		checkPodRunningReady = func() func(Gomega, *corev1.Pod) {
+			return func(g Gomega, pod *corev1.Pod) {
+				GinkgoHelper()
```
hmm I don't think you need `GinkgoHelper` for functions that receive a gomega instance from gomega

maiqueb replied:
@qinqon ^ ??

qinqon replied:
ALl right, I will remove it after CI finish so we have results

maiqueb replied:
Done; removed it.

jcaamano reviewed (commented):
Will we have integration docs for this new annotation?

maiqueb commented on contrib/kind-common:
```diff
@@ -581,3 +581,24 @@ kubectl_wait_dnsnameresolver_pods() {
   kubectl wait -n dnsnameresolver-operator --for=condition=ready pods --all --timeout=${timeout}s
 }
+
+deploy_primary_udn_nad() {
```
what do you suggest ? passt_binding ?

jcaamano replied:
I don't know. I don't know what network is this. Is there a diagram somewhere?

maiqueb replied:
This is how we install the plugin-based kubevirt binding that will be selected by users that want to connect to primary UDNs from KubeVirt. 

No docs right now.

jcaamano replied:
past_binding looks fine to me.

maiqueb replied:
went for kubevirt binding, since we'll end up replacing this binding for another - i.e. that's just an implementation detail.

What matters (IMO) is we indicate what this is for - the kubevirt integration.

maiqueb commented:
> Will we have integration docs for this new annotation?

Yes. But since we have no documentation about primary UDN, it feels very weird to add it. 

Once that exists, I will explain the annotation.

jcaamano commented:
One other question:

Would it have been theoretically possible to use the ipam claim reference in a network selection element for the default network (instead of this new annotation) leveraging `v1.multus-cni.io/default-network` annotation. I know that is out of spec but we are out of spec here anyway and we wouldn't need to define new things.

maiqueb commented:
> One other question:
> 
> Would it have been theoretically possible to use the ipam claim reference in a network selection element for the default network (instead of this new annotation) leveraging `v1.multus-cni.io/default-network` annotation. I know that is out of spec but we are out of spec here anyway and we wouldn't need to define new things.

Not sure ... how would that play along for the rest of the synthetic network selection element we are using to attach to UDN ? For UDN pods, we'd use the synthetic network-selection-element, for UDN VMs we'd use this explicitly defined network-selection-element ?... We'd merge the synthetic NSE w/ the existing one ?... 

Maybe it's me, but it sounds more complex.

jcaamano commented:
> > One other question:
> > Would it have been theoretically possible to use the ipam claim reference in a network selection element for the default network (instead of this new annotation) leveraging `v1.multus-cni.io/default-network` annotation. I know that is out of spec but we are out of spec here anyway and we wouldn't need to define new things.
> 
> Not sure ... how would that play along for the rest of the synthetic network selection element we are using to attach to UDN ? For UDN pods, we'd use the synthetic network-selection-element, for UDN VMs we'd use this explicitly defined network-selection-element ?... We'd merge the synthetic NSE w/ the existing one ?...
> 
> Maybe it's me, but it sounds more complex.

I am not sure what's the synthetic network selection element used for, but that looks like an implementation detail.

I would value more having a consistent API for our users.

maiqueb commented:
> > > One other question:
> > > Would it have been theoretically possible to use the ipam claim reference in a network selection element for the default network (instead of this new annotation) leveraging `v1.multus-cni.io/default-network` annotation. I know that is out of spec but we are out of spec here anyway and we wouldn't need to define new things.
> > 
> > 
> > Not sure ... how would that play along for the rest of the synthetic network selection element we are using to attach to UDN ? For UDN pods, we'd use the synthetic network-selection-element, for UDN VMs we'd use this explicitly defined network-selection-element ?... We'd merge the synthetic NSE w/ the existing one ?...
> > Maybe it's me, but it sounds more complex.
> 
> I am not sure what's the synthetic network selection element used for, but that looks like an implementation detail.
> 
> I would value more having a consistent API for our users.

It's how we're requesting the UDN interface attachment without persistent IPs. 

Meaning, the API is not consistent now. I don't think that can converge in this PR.

jcaamano commented:
> It's how we're requesting the UDN interface attachment without persistent IPs.

But you mean completely internal to OVN-K, we are not using it as an API to interface with any other external component. Or what does synthetic refer to in this context?

> Meaning, the API is not consistent now. I don't think that can converge in this PR.

The IPAM claim reference API could be consistent because it would be the same for secondary and primary networks.

Not for this PR, but maybe worth exploring before the final integration.

maiqueb commented:
> > It's how we're requesting the UDN interface attachment without persistent IPs.
> 
> But you mean completely internal to OVN-K, we are not using it as an API to interface with any other external component. Or what does synthetic refer to in this context?
> 
> > Meaning, the API is not consistent now. I don't think that can converge in this PR.
> 
> The IPAM claim reference API could be consistent because it would be the same for secondary and primary networks.
> 
> Not for this PR, but maybe worth exploring before the final integration.

This would maybe be an interesting idea to explore in the [enhancement proposal](https://github.com/openshift/enhancements/pull/1673).

jcaamano commented:
> Yes, but, I think that by making that API consistent, you are then making the UDN API inconsistent - if persistent IPs are used, we require a network selection element; if not, we'd just annotate the namespace.

The network selection element would not replace the namespace annotation, that is just a shortcut to default a different network in your namespace (so more akin to a shortcut for `v1.multus-cni.io/default-network`). But if you want to change the details of that network on a pod-per-pod basis, then you would have to annotate the pod with the NSE for the default network as you do today, for example, to use sr-iov on the default network with a specific pod.

maiqueb commented:
> The network selection element would not replace the namespace annotation, that is just a shortcut to default a different network in your namespace (so more akin to a shortcut for v1.multus-cni.io/default-network). But if you want to change the details of that network on a pod-per-pod basis, then you would have to annotate the pod with the NSE for the default network as you do today, for example, to use sr-iov on the default network with a specific pod.

Right now, it is impossible to refer to a primary UDN via network-selection-element. We disallow that explicitly in the code.

I think I start to see your point. I will consider this option in the [enhancement](https://github.com/openshift/enhancements/pull/1673). Thanks for raising it.

jcaamano commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -957,3 +965,7 @@ func DoesNetworkRequireTunnelIDs(netInfo NetInfo) bool {
 }
+
+func AllowsPersistentIPs(netInfo NetInfo) bool {
+	return netInfo.IsPrimaryNetwork() && netInfo.TopologyType() == types.Layer2Topology && netInfo.AllowsPersistentIPs()
```
Shouldn't this cover secondary networks as well?

maiqueb replied:
I'd rather not; I would prefer to have an helper tailored for this use: primary layer2 UDN with persistent IPs. 

I can make the name more appropriate though.

jcaamano replied:
It doesn't make sense then because what I wanted help with was: ` summarizes which networks allow persistent IPs`

maiqueb replied:
So let's put it back the way it was originally, because it is different in primary UDNs and in secondary networks (in primary UDNs it is only for layer2; in secondary networks it allows for localnet networks as well).

My bad in blindly going forward with this.

maiqueb replied:
Done.

jcaamano replied:
> So let's put it back the way it was originally, because it is different in primary UDNs and in secondary networks (in primary UDNs it is only for layer2; in secondary networks it allows for localnet networks as well).
> 
> My bad in blindly going forward with this.

That's actually the nuance that I wanted to put on the table and summarize in a single place. Wouldn't this work, for example:
```
func AllowsPersistentIPs(netInfo NetInfo) bool {
  switch {
      case netInfo.IsPrimaryNetwork():
          return netInfo.TopologyType() == types.Layer2Topology && netInfo.AllowsPersistentIPs()
  
      case netInfo.IsSecondaryNetwork():
          return (netInfo.TopologyType() == types.Layer2Topology || netInfo.TopologyType() == types.LocalnetTopology) && netInfo.AllowsPersistentIPs()
  
      default:
          return false
  }
}
```

maiqueb replied:
OK, I'll do this.

maiqueb replied:
Done; the other usages are just too weirdly scoped for me to change in the context of this PR.

jcaamano replied:
what about cluster manager? Aren't layer 2 networks reconciling ipam claims from there? This might have changed and I missed it.

maiqueb replied:
Correct; done.

qinqon commented:
@maiqueb @jcaamano the kubevirt failure is a new flake related to health checks after exposing a services to a VM 
```
[FAILED] worker1: Expose tcpServer as a service
  Unexpected error:
      <*fmt.wrapError | 0xc0005c4320>: 
      failed DialTCP: dial tcp 172.18.0.5:32102: connect: connection refused
      {
          msg: "failed DialTCP: dial tcp 172.18.0.5:32102: connect: connection refused",
          err: <*net.OpError | 0xc00057c910>{
              Op: "dial",
              Net: "tcp",
              Source: nil,
              Addr: <*net.TCPAddr | 0xc0007b58f0>{
                  IP: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 172, 18, 0, 5],
                  Port: 32102,
                  Zone: "",
              },
              Err: <*os.SyscallError | 0xc0005c4300>{
                  Syscall: "connect",
                  Err: <syscall.Errno>0x6f,
              },
          },
      }
```

jcaamano commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -939,8 +939,16 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *kapi.Pod, nInfo NetInfo, ac
 		Name:      activeNetworkNADKey[1],
 	}
 
+	if AllowsPersistentIPs(nInfo) {
```
Do we need to check if this is a primary network here?

maiqueb replied:
Why ? Isn't that checked inside the allows persistent IPs helper you've requested ?

jcaamano replied:
Right, but the annotation itself is just for primary networks and not secondary ones, right? So if we just use `AllowsPersistentIPs` we would be allowing the use of the annotation for secondary networks as well?

maiqueb replied:
Thus, IMHO, coalescing all the info about when persistent ips in a single place is counter-productive. 

... since we'll have to perform redundant checks elsewhere to ensure we're not covering more than what we want.

But I'll blindly throw it in. Done.

jcaamano replied:
The way we have the code now, which could be better but I think is an improvement over what we had, highlights that the persistent IPs functionality is the same for all networks and that the only difference is the API. I don't really understand how you see a problem with the code but at the same time be fine with the API differences.

maiqueb replied:
The UDN API differences were proposed, discussed, and accepted on an enhancement.

I personally think they do make sense, and simplify the users tasks. Hence I'm OK with it.

Still, I think this is tangential to this particular discussion.

jcaamano commented:
> @maiqueb @jcaamano the kubevirt failure is a new flake related to health checks after exposing a services to a VM, is not realted to this PR since the "post-copy" test is doing the same and is passing.

Do we need to file an issue for this one?

maiqueb commented:
> ovn-ci / e2e (kv-live-migration, noHA, shared, dualstack, SnatGW, 1br, ic-single-node-zones, 3

https://github.com/ovn-org/ovn-kubernetes/issues/4709

maiqueb commented:
The upgrade test is flaking on https://github.com/ovn-org/ovn-kubernetes/issues/4481; updated that issue.

qinqon commented:
/lgtm

The kubevirt issues were related to https://github.com/ovn-org/ovn-kubernetes/pull/4651

maiqueb commented:
Fixed the errors in the kubevirt lane, introduced after rebasing #4625 (since we were testing a dual stack UDN in an IPv4 cluster) in [this forced push](https://github.com/ovn-org/ovn-kubernetes/compare/cf318cebb7c27e7231439f8f7189d8053be28d38..46dc6a0240d0ab7a4cd99f4ba2a08c80a50efd4c).

maiqueb commented:
> /lgtm
> 
> The kubevirt issues were related to #4625

Thanks for the help !!! :rocket:
